usr.bin/script/script.c

```diff
@@ -0,0 +1,481 @@
+		warn("fork");
+		fail();
+	}
+	if (child == 0) {
+		subchild = child = fork();
+		if (child == -1) {
+			warn("fork");
+			fail();
+		}
+		if (child)
+			dooutput();
+		else
+			doshell(command);
+	}
+
+	if (!rawout)
+		(void)fclose(fscript);
+	while ((scc = read(STDIN_FILENO, ibuf, BUFSIZ)) > 0) {
+		cc = (size_t)scc;
+		if (rawout)
+			record(fscript, ibuf, cc, 'i');
+		(void)write(master, ibuf, cc);
+	}
+	childwait();
+	return EXIT_SUCCESS;
+}
+
+static void
+childwait(void)
+{
+	sigset_t set;
+
+	sigemptyset(&set);
+	sigsuspend(&set);
+}
+
+static void
+finish(int signo)
+{
+	int die, pid, status;
+
+	die = 0;
+	while ((pid = wait4(-1, &status, WNOHANG, 0)) > 0)
+		if (pid == child)
+			die = 1;
+
+	if (die)
+		done();
+}
+
+static void
+dooutput(void)
+{
+	struct itimerval value;
+	ssize_t scc;
+	size_t cc;
+	time_t tvec;
+	char obuf[BUFSIZ];
+
+	(void)close(STDIN_FILENO);
+	tvec = time(NULL);
+	if (rawout)
+		record(fscript, NULL, 0, 's');
+	else if (!quiet)
+		(void)fprintf(fscript, "Script started on %s", ctime(&tvec));
+
+	(void)signal(SIGALRM, scriptflush);
+	value.it_interval.tv_sec = SECSPERMIN / 2;
+	value.it_interval.tv_usec = 0;
+	value.it_value = value.it_interval;
+	(void)setitimer(ITIMER_REAL, &value, NULL);
+	for (;;) {
+		scc = read(master, obuf, sizeof(obuf));
+		if (scc <= 0)
+			break;
+		cc = (size_t)scc;
+		(void)write(STDOUT_FILENO, obuf, cc);
+		if (rawout)
+			record(fscript, obuf, cc, 'o');
+		else
+			(void)fwrite(obuf, 1, cc, fscript);
+		outcc += cc;
+		if (flush)
+			(void)fflush(fscript);
+	}
+	childwait();
+	exit(EXIT_SUCCESS);
+}
+
+static void
+scriptflush(int signo)
+{
+	if (outcc) {
+		(void)fflush(fscript);
+		outcc = 0;
+	}
+}
+
+static void
+doshell(const char *command)
+{
+	const char *shell;
+
+	(void)close(master);
+	(void)fclose(fscript);
+	login_tty(slave);
+	if (command == NULL) {
+		shell = getenv("SHELL");
+		if (shell == NULL)
+			shell = _PATH_KSHELL;
+		execl(shell, shell, "-i", NULL);
+		warn("execl `%s'", shell);
+	} else {
+		if (system(command) == -1)
+			warn("system `%s'", command);
+	}
+
+	fail();
+}
+
+static void
+fail(void)
+{
+
+	(void)kill(0, SIGTERM);
+	done();
+}
+
+static void
+done(void)
+{
+	time_t tvec;
+
+	if (subchild) {
+		tvec = time(NULL);
+		if (rawout)
+			record(fscript, NULL, 0, 'e');
+		else if (!quiet)
+			(void)fprintf(fscript,"\nScript done on %s",
+			    ctime(&tvec));
+		(void)fclose(fscript);
+		(void)close(master);
+	} else {
+		if (isterm)
+			(void)tcsetattr(STDIN_FILENO, TCSAFLUSH, &tt);
+		if (!quiet)
+			(void)printf("Script done, output file is %s\n", fname);
+	}
+	exit(EXIT_SUCCESS);
+}
+
+static void
+record(FILE *fp, char *buf, size_t cc, int direction)
+{
+	struct stamp stamp;
+	struct timeval tv;
+  int fd = fileno(fp);
+
+	(void)gettimeofday(&tv, NULL);
+	stamp.scr_len = cc;
+	stamp.scr_sec = tv.tv_sec;
+	stamp.scr_usec = tv.tv_usec;
+	stamp.scr_direction = direction;
+	if (write(fd, &stamp, sizeof(stamp)) != sizeof(stamp) ||
```

Comment:
Could we introduce `writev` wrapper in `libc` or better implement the syscall ?

Reply:
I have added userspace implementations of `writev()` and `readv()`.